### PR TITLE
Bug fix for drop path decay rate in swin transformer

### LINKED ISSUE #33974: 'drop_path` argument for SwinStage class is unused.
### System Info

Nothing related to system 

### Who can help?

 @amyeroberts, @qubvel

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

-

### Expected behavior

I feel like i have found some minor errors in swin transformer code.

The SwinEncoder class has layers attribute which is a ModuleList of SwinStage objects.  https://github.com/huggingface/transformers/blob/38f9f10dd9240619ea17fb6c7acb51b3bc592232/src/transformers/models/swin/modeling_swin.py#L821-L840

SwinStage class takes in `drop_path` as input argument, but in SwinStage class constructor defination, this argument is not being used anywhere.   

https://github.com/huggingface/transformers/blob/38f9f10dd9240619ea17fb6c7acb51b3bc592232/src/transformers/models/swin/modeling_swin.py#L762-L780

Probability for DropPath is picked from the config  in DropPath layer.

https://github.com/huggingface/transformers/blob/38f9f10dd9240619ea17fb6c7acb51b3bc592232/src/transformers/models/swin/modeling_swin.py#L637-L750

### PR DESCRIPTION
**Pull request by abhi-glitchhg**
# What does this PR do?

This PR fixes #33974 . 

As i had mentioned in the issue, I feel that swin transformer implementation has incorrect implementation of stochastic depth decay.

According to the official implementation, drop_prob for every SwinLayer is different. 

https://github.com/microsoft/Swin-Transformer/blob/f82860bfb5225915aca09c3227159ee9e1df874d/models/swin_transformer.py#L544

https://github.com/microsoft/Swin-Transformer/blob/f82860bfb5225915aca09c3227159ee9e1df874d/models/swin_transformer.py#L558

https://github.com/microsoft/Swin-Transformer/blob/main/models/swin_transformer.py#L397-#L408

But in transformers, we were using a constant value that is picked from the config file. I feel that implementations in transformers should be closer to the official ones.  This also applies for the SwinV2 model.  (and maybe swin2sr as well)

Please do look into this and let me know. Also i have changed the variable names as well. I am very bad at naming, so any suggestions for the argument names are welcome 😄 

 

- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. -> #33974 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@amyeroberts, @qubvel @ArthurZucker

### CODE REVIEW

File: src/transformers/models/swin/modeling_swin.py
Comment (deo-abhijit) [inline src/transformers/models/swin/modeling_swin.py:772]: self.num_layers will be undefined for this class, will do the change required

**Comment by ArthurZucker** [comment]
cc @molbap

File: src/transformers/models/swin/modeling_swin.py
Comment (molbap) [reply src/transformers/models/swin/modeling_swin.py:772]: you can pick the number of layers from the `config` object!

File: src/transformers/models/swin/modeling_swin.py
Comment (molbap) [inline src/transformers/models/swin/modeling_swin.py:842]: I think this `dpr` is undefined?

**Review by molbap — COMMENTED** [review]
Hi and thanks for the PR! I think this is a work in progress? If sofeel free to pass the PR in `draft` mode or name it [WIP], and ping me again when you want another review! I think you correctly identified this bug :) when you're done, you can run `make fixup` to run the linter and make `check doe quality` happy in the CI

**Comment by abhi-glitchhg** [comment]
Hi, thanks for the reply. I was waiting for confirmation if this was indeed a bug or intended behaviour. I will fix the lint and other errors.  
Thanks.
abhijit

**Comment by abhi-glitchhg** [comment]
In the #33974 , @ArthurZucker mentioned the messy initialisation of swinlayer class, i would not like to touch it in this pr. 

Personally i think  the initialisation looks good. But if you think we should make it simpler, i m happy.

**Comment by abhi-glitchhg** [comment]
```
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::DeiT2RobertaModelTest::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 21]) != (13, 20)
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::ViT2BertModelTest::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 21]) != (13, 20)
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::Swin2BartModelTest::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 21]) != (13, 20)
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::ViT2TrOCR::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 21]) != (13, 20)
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::LayoutLMv32TrOCR::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 28]) != (13, 20)
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::VIT2GPT2Test::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 27]) != (13, 20)
FAILED tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::Donut2GPT2Test::test_encoder_decoder_model_generate - AssertionError: torch.Size([13, 27]) != (13, 20)
```

Hmm, these tests are failing in CI.

File: src/transformers/models/glm/modeling_glm.py
Comment (deo-abhijit) [inline src/transformers/models/glm/modeling_glm.py:55]:
```suggestion
```

File: src/transformers/models/glm/modeling_glm.py
Comment (abhi-glitchhg) [inline src/transformers/models/glm/modeling_glm.py:53]:
```suggestion
```

**Comment by abhi-glitchhg** [comment]
The CI is green. Yey! ig now the pr is ready for review @molbap 

Btw i really loved the infra for CI/CD, tests run fast! Linting tools work amazingly!  I also recently watched @ArthurZucker's  pytorch conference talk and now i **really** understand the pain points mentioned in the video. 

Thank you guys for maintaining such a high-impact library!

File: src/transformers/models/swin/modeling_tf_swin.py
Comment (abhi-glitchhg) [inline src/transformers/models/swin/modeling_tf_swin.py:752]: this is all because of ruff linter

File: src/transformers/models/swinv2/modeling_swinv2.py
Comment (molbap) [inline src/transformers/models/swinv2/modeling_swinv2.py:688]: As mentioned in the other linked PR I think it'd be nice to refactor the init of all of these layers - the reason is that we're trying to have not too many args, and here we're adding yet another one. `input_resolution`, `num_heads` etc can all be derived from `config` and `layer_idx` (albeit through `grid_size` for `input_resolution`). We should modify this in a follow-up PR, or in this one if you feel up to it!

**Review by molbap — COMMENTED** [review]
Thanks, it's cleaner! left a couple comments, let me know what you think

File: src/transformers/models/swinv2/modeling_swinv2.py
Comment (abhi-glitchhg) [reply src/transformers/models/swinv2/modeling_swinv2.py:688]: happy to do it follow-up pr

Comment (molbap) [reply src/transformers/models/swinv2/modeling_swinv2.py:688]: sounds good then!

**Review by molbap — APPROVED** [review]
LGTM - revamp of layers init to depend on (config, layer_idx) TBD in a follow-up PR! cc @ArthurZucker for final review

**Review by ArthurZucker — APPROVED** [review]
Nice and simple! Thanks 🤗

**Pull request merged**
